File: DAA/FractionalKnapsack.java
import java.util.*;

class Item implements Comparable<Item> {
    double weight;
    double value;
    double valuePerWeight;

    public Item(double weight, double value) {
        this.weight = weight;
        this.value = value;
        this.valuePerWeight = value / weight;
    }

    @Override
    public int compareTo(Item other) {
        // Sort in descending order of value per weight ratio
        return Double.compare(other.valuePerWeight, this.valuePerWeight);
    }
}

public class FractionalKnapsack {

    public static double fractionalKnapsack(double[] weights, double[] values, double capacity) {
        int n = weights.length;
        Item[] items = new Item[n];

        // Create Item objects and calculate value per weight ratio
        for (int i = 0; i < n; i++) {
            items[i] = new Item(weights[i], values[i]);
        }

        // Sort items based on value per weight ratio
        Arrays.sort(items);

        double totalValue = 0;
        double remainingCapacity = capacity;

        // Iterate through sorted items
        for (Item item : items) {
            if (remainingCapacity >= item.weight) {
                // If capacity is enough, add the whole item
                totalValue += item.value;
                remainingCapacity -= item.weight;
            } else {
                // If not, add a fraction of the item
                totalValue += (item.valuePerWeight * remainingCapacity);
                break;
            }
        }

        return totalValue;
    }

    public static void main(String[] args) {
        double[] weights = {10, 20, 30, 40, 50 };
        double[] values = {20, 30, 66, 40, 60 };
        double capacity = 100;

        double maxTotalValue = fractionalKnapsack(weights, values, capacity);
        System.out.println("Maximum value: " + maxTotalValue);
    }
}

File: DAA/FractionalKnapsack.c
#include<stdio.h>

int main()
{
	int n,i,j,temp;
	printf("Enter length of array:");
	scanf("%d",&n);
	float weight[n];
	float value[n];
	float ratio[n];
	float capacity;
	float x[n];
	
	printf("Enter capacity of the knapsack:");
	scanf("%f",&capacity);
	
	for(i=0;i<n;i++){
		printf("Enter weight of %d item: ",i);
		scanf("%f",&weight[i]);
		printf("Enter value of %d item: ",i);
		scanf("%f",&value[i]);
		
		x[i] = 0.0;
		ratio[i] = value[i] / weight[i];
	}
	
	for(i=0;i<n;i++) {
		for(j=i+1;j<n;j++) {
			if(ratio[i] < ratio[j]) {
				temp = ratio[j];
				ratio[j] = ratio[i];
				ratio[i] = temp;
				
				temp = weight[j];
				weight[j] = weight[i];
				weight[i] = temp;
				
				temp = value[j];
				value[j] = value[i];
				value[i] = temp;
			}
		}
	}
	
	float u = capacity;
	float tp = 0.0;
	
	for(i=0;i<n;i++) {
		if(weight[i] > u) {
			break;
		}
		else {
			x[i] = 1.0;
			tp += value[i];
			u -= weight[i];
		}
	}
	
	if(i<n) {
		x[i] = u / weight[i];
	}
	
	tp += x[i] * value[i];
	
	printf("\nWeight:\t");
	for(i=0;i<n;i++) {
		printf("Weight: %1.2f\t",weight[i]);
			printf("Value: %1.2f\t",value[i]);
				printf("Ratio: %1.2f\t",x[i]);
	}
	
	printf("\nMaximum Profit: %f",tp);
	
	return 0;
}

